src/v/iceberg/rest_client/catalog_client.cc

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "iceberg/rest_client/catalog_client.h"
+
+#include "bytes/iobuf_parser.h"
+#include "http/request_builder.h"
+#include "http/utils.h"
+#include "iceberg/rest_client/parsers.h"
+
+#include <seastar/coroutine/as_future.hh>
+
+namespace {
+
+std::optional<std::string_view>
+trim_slashes(std::optional<std::string_view> s) {
+    if (!s.has_value()) {
+        return s;
+    }
+
+    auto v = s.value();
+    if (v.starts_with("/")) {
+        v.remove_prefix(1);
+    }
+
+    if (v.ends_with("/")) {
+        v.remove_suffix(1);
+    }
+
+    return v;
+}
+
+} // namespace
+
+namespace iceberg::rest_client {
+
+expected<json::Document> parse_json(iobuf&& raw_response) {
+    iobuf_parser p{std::move(raw_response)};
+    auto raw_json = p.read_string(p.bytes_left());
+
+    json::Document doc;
+    doc.Parse(raw_json);
+
+    if (doc.HasParseError()) {
+        return tl::unexpected(json_parse_error{
+          .context = "parse_json",
+          .error = GetParseError_En(doc.GetParseError())});
+    }
+
+    return doc;
+}
+
+catalog_client::catalog_client(
+  http::abstract_client& client,
+  std::string_view endpoint,
+  credentials credentials,
+  std::optional<std::string_view> base_url,
+  std::optional<std::string_view> prefix,
+  std::string_view api_version,
+  std::optional<oauth_token> token,
+  std::unique_ptr<retry_policy> retry_policy)
+  : _client{client}
+  , _endpoint{endpoint}
+  , _credentials{std::move(credentials)}
+  , _base_url{trim_slashes(base_url).value_or("")}
```

Comment:
When does a `base_url` have a leading slash? Is this actually a `base_path`? Also feels like you could simplify `trim_slashes` by doing `absl::StripSuffix(base_url.value_or(""), "/")`

Reply:
changed to `base_path`, I also extracted a `path_components` struct as path computation was getting complex and hard to test. 

I will try to add named types for the three path components as it is easy to mix up adjacend `ss::sstring`s

Reply:
added `named_type`s instead of `ss::sstring`s for path components